File: guides/micronaut-microservices-distributed-tracing-zipkin/bookrecommendation/kotlin/src/test/kotlin/example/micronaut/BookInventoryClientStub.kt
package example.micronaut

import io.micronaut.context.annotation.Requires
import io.micronaut.context.env.Environment
import io.micronaut.retry.annotation.Fallback
import org.reactivestreams.Publisher
import reactor.core.publisher.Mono
import jakarta.inject.Singleton
import jakarta.validation.constraints.NotBlank

@Requires(env = arrayOf(Environment.TEST))
@Fallback
@Singleton
open class BookInventoryClientStub : BookInventoryOperations {

    override fun stock(@NotBlank isbn: String): Mono<Boolean> {
        if (isbn == "1491950358") {
            return Mono.just(java.lang.Boolean.TRUE)

        } else if (isbn == "1680502395") {
            return Mono.just(java.lang.Boolean.FALSE)
        }
        return Mono.empty()
    }
}

File: guides/micronaut-microservices-distributed-tracing-zipkin-opentelemetry/bookrecommendation/python/tests/example/micronaut/book_inventory_client_stub.py
from typing import Annotated

from jakarta.inject import Singleton
from jakarta.validation.constraints import NotBlank
from micronaut.context.annotation import Requires
from micronaut.core.async_.annotation import SingleResult
from micronaut.retry.annotation import Fallback
from org.reactivestreams import Publisher
from reactor.core.publisher import Mono

from example.micronaut.book_inventory_operations import BookInventoryOperations


@Requires(env="test")
@Fallback
@Singleton
class BookInventoryClientStub(BookInventoryOperations):

    @SingleResult
    def stock(self, isbn: Annotated[str, NotBlank]) -> Publisher[bool]:
        if isbn == "1491950358":
            return Mono.just(True)
        if isbn == "1680502395":
            return Mono.just(False)
        return Mono.empty()
